editor/src/core/workers/parser-printer/uid-fix.ts

```diff
@@ -1,284 +1,569 @@
-        const uidFix = newToOldUidMapping[highlightBound.uid]
-        if (uidFix != null) {
-          return {
-            ...highlightBound,
-            uid: uidFix.oldUID,
-          }
-        } else {
-          return highlightBound
-        }
-      },
-    )
 
-    return {
-      ...newParsed,
-      topLevelElements: fixedTopLevelElements,
-      highlightBounds: fixedHighlightBounds,
-    }
-  } else {
-    // we found no uid mismatch or we found too many mismatched UIDs and so we bail out
-    return newParsed
+  // This needs to be corrected as things may have moved around.
+  const fixedHighlightBounds: HighlightBoundsForUids = {
+    ...newParsed.highlightBounds,
+  }
+  for (const { oldUID, newUID } of fixUIDsState.mappings) {
+    const bounds = fixedHighlightBounds[oldUID]
+    delete fixedHighlightBounds[oldUID]
+    fixedHighlightBounds[newUID] = bounds
+  }
+
+  // Return the result.
+  return {
+    ...newParsed,
+    topLevelElements: fixedTopLevelElements,
+    combinedTopLevelArbitraryBlock: fixedCombinedTopLevelArbitraryBlock,
+    highlightBounds: fixedHighlightBounds,
   }
 }
 
-type OnElement = (
+function updateUID<T>(
+  uidOptic: Optic<T, string>,
   oldUID: string,
-  newUID: string,
-  oldPathToRestore: StaticElementPathPart,
-  newElementPath: StaticElementPathPart,
-) => void
-
-function zipTopLevelElements(
-  firstTopLevelElements: Array<TopLevelElement>,
-  secondTopLevelElements: Array<TopLevelElement>,
-  onElement: OnElement,
-): boolean {
-  const firstComponents = getComponentsFromTopLevelElements(firstTopLevelElements)
-  const secondComponents = getComponentsFromTopLevelElements(secondTopLevelElements)
-
-  const lengthsMatchAtThisLevel = firstComponents.length === secondComponents.length
-  let numberOfElementsMatch = lengthsMatchAtThisLevel
-
-  firstComponents.forEach((firstComponent, index) => {
-    if (secondComponents.length > index) {
-      const secondComponent = secondComponents[index]
-
-      const lengthsOfChildrenMatch = walkElementChildren(
-        EP.emptyElementPathPart,
-        [firstComponent.rootElement],
-        [secondComponent.rootElement],
-        onElement,
-      )
-      numberOfElementsMatch = numberOfElementsMatch && lengthsOfChildrenMatch
+  fixUIDsState: FixUIDsState,
+  baseValue: T,
+): T {
+  const newUID = unsafeGet(uidOptic, baseValue)
+  let uidToUse: string = oldUID
+  if (oldUID === newUID) {
+    // Old one is the same as the new one, so everything is great.
+    uidToUse = newUID
+  } else if (fixUIDsState.mutableAllNewUIDs.has(oldUID)) {
+    // The old uid is already present somewhere, so using it will introduce a duplicate.
+    uidToUse = newUID
+  } else {
+    // The uids have changed:
+    // - Remove the new one from the set, as now it will be unused.
+    // - Add the old one to the set, as it will become used.
+    // - Return the old one.
+    // One assumption here is that there is no duplicate uids in the new elements.
```

Comment:
Hmmm can we do anything about this assumption so that we don't have to consider it an assumption? This sounds like something that would happen after a `cmd`+`d` duplication.

Reply:
Currently we only utilise this during a parse and the parse should have resulted in there being no duplicates. It seemed worth mentioning just in case some future changes meant that this assumption resulted in breakage.